### ISSUE TITLE
Inquiry About Method’s Runtime Performance

### ISSUE THREAD

**Issue by Snow2022jlu** [zikai1/OAReg#1]
Dear Author,

Hello! I hope this message finds you well. I’m reproducing the method described in your paper and have encountered a performance discrepancy that I would greatly appreciate your guidance on.

In Table 2 of your paper, you report that your method completes non-rigid registration of two models in just 13.646 seconds. However, when I ran your publicly released code under the same settings (using models with around 10,000 points), a single run took approximately 100 seconds.

My current understanding is as follows:

The line

`loss = local_linear_reconstruction(x_canonical, x_deformed, n_neighbors=30)`

appears to use barycenter_kneighbors_graph and barycenter_weights, which are CPU‑based (NumPy/SciPy/scikit‑learn), while only subsequent tensor operations run on the GPU. I suspect that in your experiments you may have employed an all‑GPU implementation—for example, using `、torch.cdist+ topk`  for KNN and torch.linalg.solve for batched solutions—but the released repository currently contains only the CPU version.

My questions are:

In your paper’s experiments, did you use a pure GPU implementation of these components?

If so, would you be willing to share or point me to guidance on converting the open‑source version to a GPU‑accelerated one?

Could you consider including a full GPU implementation in future repository updates to facilitate reproducibility?

For reference, here is my reproduction setup:

OS: Ubuntu 20.04

GPU: NVIDIA RTX 3090, CUDA 11.3

CPU: Intel® Xeon® Silver 4215R @ 3.20 GHz

Thank you very much for sharing your excellent work. I look forward to any advice or pointers you can provide!

**Comment by zikai1** [comment]
Hi, thanks for your interest. The point size of your model may be excessive and you could try downsampling operations such as voxel downsampling to reduce the point cloud size and accelerate registration. Additionally, you can moderately decrease the value of `n_neighbors` used in local linear reconstruction, as well as the number of optimization iterations. Typically, lower `n_neighbors` values yield faster speeds but slightly reduced quality.

Yes, I agree that a more effective approach would be to use a GPU to accelerate the computation of barycenter_kneighbors_graph and barycenter_weights. In my experiments, I focused primarily on measuring the registration time, excluding the lengthy data loading process, as including this would bias the evaluation of the registration algorithm itself.

**Comment by zikai1** [comment]
> Dear Author,
> 
> Hello! I hope this message finds you well. I’m reproducing the method described in your paper and have encountered a performance discrepancy that I would greatly appreciate your guidance on.
> 
> In Table 2 of your paper, you report that your method completes non-rigid registration of two models in just 13.646 seconds. However, when I ran your publicly released code under the same settings (using models with around 10,000 points), a single run took approximately 100 seconds.
> 
> My current understanding is as follows:
> 
> The line
> 
> `loss = local_linear_reconstruction(x_canonical, x_deformed, n_neighbors=30)`
> 
> appears to use barycenter_kneighbors_graph and barycenter_weights, which are CPU‑based (NumPy/SciPy/scikit‑learn), while only subsequent tensor operations run on the GPU. I suspect that in your experiments you may have employed an all‑GPU implementation—for example, using `、torch.cdist+ topk` for KNN and torch.linalg.solve for batched solutions—but the released repository currently contains only the CPU version.
> 
> My questions are:
> 
> In your paper’s experiments, did you use a pure GPU implementation of these components?
> 
> If so, would you be willing to share or point me to guidance on converting the open‑source version to a GPU‑accelerated one?
> 
> Could you consider including a full GPU implementation in future repository updates to facilitate reproducibility?
> 
> For reference, here is my reproduction setup:
> 
> OS: Ubuntu 20.04
> 
> GPU: NVIDIA RTX 3090, CUDA 11.3
> 
> CPU: Intel® Xeon® Silver 4215R @ 3.20 GHz
> 
> Thank you very much for sharing your excellent work. I look forward to any advice or pointers you can provide!

Hi, thanks for your interest. The point size of your model may be excessive and you could try downsampling operations such as voxel downsampling to reduce the point cloud size and accelerate registration. Additionally, you can moderately decrease the value of n_neighbors used in local linear reconstruction, as well as the number of optimization iterations. Typically, lower n_neighbors values yield faster speeds but slightly reduced quality.

Yes, I agree that a more effective approach would be to use a GPU to accelerate the computation of barycenter_kneighbors_graph and barycenter_weights. In my experiments, I focused primarily on measuring the registration time, excluding the lengthy data loading process, as including this would bias the evaluation of the registration algorithm itself.

**Comment by Snow2022jlu** [comment]
Thank you very much for your prompt and detailed response to my question!

Your suggestions regarding point cloud downsampling, the number of neighbors (n_neighbors), and optimization iterations were extremely helpful and had a significant impact on my reproduction efforts. That said, moving the computation of local_linear_reconstruction to the GPU proved to be even more effective in my case—I implemented a GPU-based version of the algorithm and saw a substantial speedup, reducing the runtime from 100 seconds to around 30 seconds.

I truly appreciate your patience and generosity in taking the time to respond despite your busy schedule, and I sincerely look forward to more of your outstanding research in the future!

**Comment by Snow2022jlu** [comment]
> > Dear Author,
> > Hello! I hope this message finds you well. I’m reproducing the method described in your paper and have encountered a performance discrepancy that I would greatly appreciate your guidance on.
> > In Table 2 of your paper, you report that your method completes non-rigid registration of two models in just 13.646 seconds. However, when I ran your publicly released code under the same settings (using models with around 10,000 points), a single run took approximately 100 seconds.
> > My current understanding is as follows:
> > The line
> > `loss = local_linear_reconstruction(x_canonical, x_deformed, n_neighbors=30)`
> > appears to use barycenter_kneighbors_graph and barycenter_weights, which are CPU‑based (NumPy/SciPy/scikit‑learn), while only subsequent tensor operations run on the GPU. I suspect that in your experiments you may have employed an all‑GPU implementation—for example, using `、torch.cdist+ topk` for KNN and torch.linalg.solve for batched solutions—but the released repository currently contains only the CPU version.
> > My questions are:
> > In your paper’s experiments, did you use a pure GPU implementation of these components?
> > If so, would you be willing to share or point me to guidance on converting the open‑source version to a GPU‑accelerated one?
> > Could you consider including a full GPU implementation in future repository updates to facilitate reproducibility?
> > For reference, here is my reproduction setup:
> > OS: Ubuntu 20.04
> > GPU: NVIDIA RTX 3090, CUDA 11.3
> > CPU: Intel® Xeon® Silver 4215R @ 3.20 GHz
> > Thank you very much for sharing your excellent work. I look forward to any advice or pointers you can provide!
> 
> Hi, thanks for your interest. The point size of your model may be excessive and you could try downsampling operations such as voxel downsampling to reduce the point cloud size and accelerate registration. Additionally, you can moderately decrease the value of n_neighbors used in local linear reconstruction, as well as the number of optimization iterations. Typically, lower n_neighbors values yield faster speeds but slightly reduced quality.
> 
> Yes, I agree that a more effective approach would be to use a GPU to accelerate the computation of barycenter_kneighbors_graph and barycenter_weights. In my experiments, I focused primarily on measuring the registration time, excluding the lengthy data loading process, as including this would bias the evaluation of the registration algorithm itself.

Thank you very much for your prompt and detailed response to my question!

Your suggestions regarding point cloud downsampling, the number of neighbors (n_neighbors), and optimization iterations were extremely helpful and had a significant impact on my reproduction efforts. That said, moving the computation of local_linear_reconstruction to the GPU proved to be even more effective in my case—I implemented a GPU-based version of the algorithm and saw a substantial speedup, reducing the runtime from 100 seconds to around 30 seconds.

I truly appreciate your patience and generosity in taking the time to respond despite your busy schedule, and I sincerely look forward to more of your outstanding research in the future!

**Comment by zikai1** [comment]
Thank you very much for boosting the registration speed by a GPU implementation. Hope you can further push forward the development of non-rigid shape registration :beers: . 

I will close this issue now but it is free to reopen it if you have any other questions.
